### PR TITLE
feat(expo): add config plugin for Android native initialization (MSDK-404)

### PR DESCRIPTION
**Pull request by msouza-attentive**
## Summary

Expo apps using [Continuous Native Generation](https://docs.expo.dev/workflow/continuous-native-generation/) regenerate `android/` on every `expo prebuild`, which wipes the manual `AttentiveSdk.initialize()` call this SDK requires in `Application.onCreate()` ([README § Android init](https://github.com/attentive-mobile/attentive-react-native-sdk/blob/main/README.md#android--initialize-from-native-code)). Until now that made the SDK effectively incompatible with CNG-based Expo apps.

This PR adds an **Expo config plugin** that injects the Android native initialization into the generated `MainApplication.kt` during prebuild, so it survives every regeneration:

```json
"plugins": [
  ["@attentive-mobile/attentive-react-native-sdk", { "domain": "myshop", "mode": "production" }]
]
```

`domain` is required (validated with an actionable error, including when the plugin is listed as a bare string); `mode` is optional and defaults to `production`.

## How it works

- `withMainApplication` + `mergeContents` insert two tagged `@generated` blocks (imports after the `package` line, init after `super.onCreate()`). Re-runs are idempotent; prop changes update the block in place; the domain is emitted through a Kotlin-literal escaper so no value can produce invalid generated code.
- **Manual-integration safety:** detection strips our tagged blocks first, so Attentive code the app owns is recognized whether or not plugin-managed blocks exist — the plugin skips and warns rather than double-initializing (it never deletes app-owned code).
- **Kotlin-only by design**: requires Expo SDK 50+ (the Kotlin-template era). On older Java templates the plugin fails prebuild with an explicit error instead of generating broken code.
- **iOS is deliberately untouched** — iOS initializes from TypeScript, so the plugin registers no iOS mods.
- **Module resolution** is centralized in `plugin/src/expoConfigPlugins.ts` and hardened for real-world layouts:
  - `expo/config-plugins` resolves by bare require first (standard layouts; pnpm resolves through our optional peer), falling back to resolving from `process.cwd()` — the app project root during prebuild — which covers workspace monorepos where npm hoists this SDK to the workspace root while `expo` stays in the app's own `node_modules`.
  - `mergeContents` uses the public `CodeGenerator` export on Expo SDK 54+, falling back on SDK 50–53 (which don't re-export it) to resolving `@expo/config-plugins`' internal `generateCode` module *through the app's expo package* — a declared dependency edge, valid under every package manager including pnpm's strict layout, against SDK versions frozen on npm.
- Packaging: `plugin/src` (TS) compiles to `plugin/build` (CJS) during `prepare` (so git-URL / commit-pin installs build from source, not just registry tarballs); `app.plugin.js` + `plugin/build` ship via the `files` allowlist. Bare-RN consumers see zero behavior change.

## Public API impact

Additive only. New config-plugin entry point (`app.plugin.js`); no changes to the existing TypeScript API, TurboModule spec, or either native module (the only `src/` change is JSDoc). New **optional** peerDependency `expo >=50.0.0` (npm/yarn/pnpm will not install or require it for non-Expo apps); new devDependencies `expo` and `@types/node` (repo-only, for types/build).

## Verification

- **Unit**: 14 plugin tests against the pristine `expo-template-bare-minimum@57.0.8` `MainApplication.kt` fixture — injection anchors, mode mapping (incl. omitted-mode default), idempotency, in-place prop updates with duplicate-block guard, manual-integration skip (both pure-manual and manual-alongside-generated-blocks), Java throw, prop validation errors, Kotlin-literal escaping, a full-output snapshot, and a masked-export test pinning the pre-SDK-54 fallback. Full suite: 51/51. Plugin code is covered by the existing CI `lint`/`typecheck`/`unit-test` jobs; typecheck verified under both tsc 4.9 (repo) and 7.0 (current editors).
- **Packaging**: `npm pack` ships `app.plugin.js` + `plugin/build`; the tarball installs and resolves by package name from a consumer `app.json`.
- **Android E2E across the support range**:
  - **Expo SDK 50** (`expo@50.0.21`, where `CodeGenerator` is confirmed absent): prebuild succeeds through the deep-resolve fallback; Kotlin template confirmed at the exact floor of the range.
  - **Expo SDK 56, workspace monorepo** (SDK hoisted to workspace root, expo in the app workspace — the layout that crashes without the cwd fallback): prebuild succeeds and injects correctly; also re-run without `--clean` → byte-identical; domain change → block updated in place; **gradle compile + runtime `identify` verified on device**; manual-code-alongside-blocks → skip + warn, file untouched.
  - **Expo SDK 57** scratch app: prebuild output byte-identical to the unit-test snapshot (same content hashes).
- **iOS E2E**: `expo prebuild -p ios --clean` with the plugin listed succeeds; zero Attentive references in the generated `ios/`.
- **Old vs new architecture**: N/A — no native module or bridge changes; the plugin runs at prebuild time in Node, not in the app.

## Rollback

- **Host apps**: remove the plugin entry from `app.json` and re-run prebuild — the `@generated` blocks disappear with the next regeneration; no persistent state anywhere.
- **This repo**: revert the branch commits; nothing else depends on `plugin/`.

## Docs

README gained an "Expo apps — Config plugin (Android)" section (+ a CNG warning in the manual-init section); AGENTS.md branches Expo-CNG vs bare-RN in its detection and init steps; CLAUDE.md documents the `plugin/` layout and its import constraint; `initialize()` JSDoc updated.

## Known limitations (follow-ups, not blockers)

- Builder options beyond `domain`/`mode` (`notificationIconId`, `skipFatigueOnCreatives`, `logLevel`) are not yet exposed as plugin props.
- Push-notification setup on Android (manifest/FCM/service changes per `docs/PUSH_NOTIFICATIONS_INTEGRATION.md`) is still manual native editing and therefore not yet CNG-compatible — same class of problem this plugin solves for init, deserving its own ticket.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: plugin/src/withAndroid.ts [hunk 1]
```diff
@@ -0,0 +1,125 @@
+import {
+  withMainApplication,
+  CodeGenerator,
```
[the remark this answers is not in the document]
Comment (msouza-attentive) [reply plugin/src/withAndroid.ts]: Confirmed — verified against the `sdk-50`, `sdk-53`, and `sdk-54` tags of expo/expo: the `CodeGenerator` namespace export first appears in SDK 54, so this import was `undefined` on SDK 50–53 and prebuild would crash inside the advertised `>=50` range. Our E2E runs were on SDK 56/57, which is why it slipped through.

Fixed in dafaafb, keeping the SDK 50 floor: the plugin now tries the public `CodeGenerator.mergeContents` export first (SDK 54+) and falls back to resolving `@expo/config-plugins/build/utils/generateCode` **through the app's `expo` package** via `createRequire(require.resolve('expo/config-plugins'))` — `expo` declares `@expo/config-plugins` as a dependency, so that resolution edge is valid under every package manager including pnpm's strict layout (a bare deep require from this package would be a phantom dependency there).

Safety of the fallback path, verified against the published artifacts: SDK 50's `@expo/config-plugins@7.8.4` has no `exports` map and ships `build/utils/generateCode.js` exporting `mergeContents`; SDK 54+'s `exports` map explicitly allows `./build/*`. Since the fallback only ever executes against SDK 50–53 — frozen, immutable versions on npm — the internal path cannot move for the versions that use it, and all future SDKs take the public branch.

Added a unit test that masks `CodeGenerator` from the mocked module to pin the fallback branch (`plugin/src/__tests__/withAndroid.test.ts`).

**Comment by msouza-attentive** [comment]
Two additional rounds of E2E testing, one new fix (8b58aa2):

**Workspace-monorepo resolution fix.** Testing against an npm-workspaces monorepo (app in a workspace subfolder) surfaced a layout the plugin didn't survive: npm hoisted this SDK to the workspace **root** `node_modules` while `expo` stayed in the app workspace's own `node_modules` — off the walk-up path from our package — so the plugin's bare `require('expo/config-plugins')` threw `Cannot find module` and prebuild crashed before injection. This predates the mergeContents work (the top-level import itself failed). Fix: resolution is centralized in `plugin/src/expoConfigPlugins.ts` — bare require first (standard layouts; pnpm resolves through the optional peer), falling back to resolving from `process.cwd()`, which `@expo/cli` sets to the app project root during prebuild. Verified against the exact broken layout: prebuild now succeeds and injects correctly with the SDK at the monorepo root and `expo@56` in the app workspace.

**Expo SDK 50 E2E (closes the caveat from the mergeContents thread).** Fresh app on `expo@50.0.21`, where I confirmed `require('expo/config-plugins').CodeGenerator === undefined` — so the prebuild could only succeed through the pre-SDK-54 deep-resolve fallback. It did: `MainApplication.kt` (Kotlin, confirming the template assumption at the exact floor of our supported range) generated with both `@generated` blocks, correct domain/mode, init immediately after `super.onCreate()`. The SDK 50–53 branch is now verified end-to-end, not just unit-tested.

Coverage is now: unit 45/45 (incl. a masked-export test for the pre-54 branch) · prebuild E2E on Expo SDK 50, 56 (monorepo + standalone), 57 · gradle compile + runtime identify on SDK 56 · iOS untouched.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: app.plugin.js [hunk 2]
```diff
@@ -0,0 +1 @@
+module.exports = require('./plugin/build')
```
[the remark this answers is not in the document]
Comment (msouza-attentive) [reply app.plugin.js:1]: Not an issue in practice — Expo's plugin loader unwraps this shape explicitly. `resolveConfigPluginExport` in `@expo/config-plugins/build/utils/plugin-resolver.js` does:

```js
if (plugin.default != null) {
  plugin = plugin.default;
}
```

(verified in the SDK 50 copy — the oldest version this PR supports — and present through current). `module.exports = require('./plugin/build')` with a TS `export default` is also the exact pattern Expo's own plugin docs template uses. Empirically: the E2E prebuilds on Expo SDK 50, 56, and 57 documented on this PR all resolved the plugin by package name through this entry point.

File: plugin/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+const packageName = '@attentive-mobile/attentive-react-native-sdk'
+
+export type AttentivePluginProps = {
+  domain: string
+  mode: 'debug' | 'production'
+}
+
+const withAttentive: ConfigPlugin<AttentivePluginProps> = (config, props) => {
+  return withAndroid(config, props)
+}
```
[the remark this answers is not in the document]
Comment (msouza-attentive) [reply plugin/src/index.ts]: Valid — fixed in 0d89f56. `withAttentive` now validates props at config-evaluation time: using the plugin as a bare string (undefined props) or without `domain` throws an actionable error that shows the expected `app.json` shape, instead of a TypeError inside codegen. `mode` is now optional (validated when present, defaults to `production`). Unit tests cover all three failure shapes. Note this validation deliberately runs in the plugin body — not the Android mod — so a misconfigured plugin fails iOS prebuilds too rather than shipping an app whose Android SDK silently never initializes.

File: plugin/src/withAndroid.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+// True when MainApplication references the Attentive SDK outside a block this
+// plugin generated — i.e. the app integrated manually. Blocks we generated
+// don't count: mergeContents replaces those on its own when props change.
+function hasManualAttentiveInit(contents: string): boolean {
+  return (
+    !contents.includes(GENERATED_BLOCK_MARKER) &&
+    (contents.includes('AttentiveConfig') || contents.includes('AttentiveSdk'))
+  )
+}
```
[the remark this answers is not in the document]
Comment (msouza-attentive) [reply plugin/src/withAndroid.ts:86]: Valid — fixed in 0d89f56. `hasManualAttentiveInit` now strips our `@generated` blocks before searching, so Attentive references in app-owned code are detected even when plugin-managed blocks exist. The mixed case (our blocks + manual init added later) skips injection and warns about the potential double-initialize, same conservative behavior as the pure-manual case. Test added.

File: plugin/src/withAndroid.ts [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+function generateInitCode(props: AttentivePluginProps): string {
+  const lines = [
+    'val attentiveConfig = AttentiveConfig.Builder()',
+    `  .applicationContext(this)`,
+    `  .domain("${props.domain}")`,
+    `  .mode(${generateMode(props)})`,
+    `  .build()`,
+    `AttentiveSdk.initialize(attentiveConfig)`,
+  ]
```
[the remark this answers is not in the document]
Comment (msouza-attentive) [reply plugin/src/withAndroid.ts:54]: Valid — fixed in 0d89f56 with the escaping approach: the domain is emitted through `kotlinStringLiteral()`, which is `JSON.stringify` plus escaping `$` (Kotlin treats `$` as template interpolation inside string literals, which JSON escaping alone would miss). Quotes/backslashes/control characters in the value now produce a valid Kotlin literal instead of failing the Android build. Test added covering quotes, backslashes, and dollar signs.

File: plugin/src/withAndroid.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+// that path *through the app's expo package*: expo declares
+// @expo/config-plugins as a dependency, so the edge is valid under every
+// package manager, including pnpm's strict layout. SDK 50–53 are frozen on
+// npm, so the internal path cannot move for the only versions that use it.
+function resolveMergeContents(): MergeContentsFn {
+  if (configPlugins.CodeGenerator?.mergeContents) {
+    return configPlugins.CodeGenerator.mergeContents
+  }
+  return requireFromExpoConfigPlugins(
+    '@expo/config-plugins/build/utils/generateCode'
```
Comment (adelag-attentive) [inline plugin/src/withAndroid.ts:24]: **[1] Kotlin string injection via unescaped `domain` prop.**

`props.domain` comes from untyped `app.json` and is interpolated directly into generated Kotlin with no escaping. A benign typo like `domain: 'my"shop'` produces `.domain("my"shop")` — prebuild succeeds, gradle fails with a confusing compile error inside a `@generated (DO NOT MODIFY)` block. A backslash or newline in the value is worse.

**Fix:** either validate against `/^[A-Za-z0-9._-]+$/` and throw early with a clear message, or run the value through `JSON.stringify()` to get properly-escaped Kotlin string literal semantics (Kotlin basic string literals are a superset of JSON escapes for the printable subset).

File: package.json [hunk 7]
```diff
@@ -29,14 +29,17 @@
...
   "scripts": {
     "test": "jest",
-    "typecheck": "tsc --noEmit",
-    "lint": "eslint \"src/**/*.{js,ts,tsx}\"",
+    "typecheck": "tsc --noEmit && tsc --noEmit --project plugin",
+    "lint": "eslint \"src/**/*.{js,ts,tsx}\" \"plugin/src/**/*.ts\"",
+    "build:plugin": "tsc --project plugin",
     "build": "bob build",
-    "prepack": "bob build",
+    "prepack": "bob build && npm run build:plugin",
```
Comment (adelag-attentive) [inline package.json]: **[3] `plugin/build/` missing when consumers install from a git URL.**

`plugin/build/` is `.gitignore`d and produced only by `prepack`. npm runs `prepack` when tarballing (`npm publish`) but **not** on `npm install <git-url>` — for git deps npm runs `prepare` and `install` scripts only. Consumers who pin the SDK to a commit (pre-release testing, private forks, monorepo path deps) will see Expo fail with `Cannot find module './plugin/build'` from `app.plugin.js`.

**Fix:** add `"prepare": "bob build && npm run build:plugin"` (or hoist the build out of `prepack`) so git installs also produce `plugin/build/`.

File: plugin/src/withAndroid.ts (hunk 6 above)
Comment (msouza-attentive) [reply plugin/src/withAndroid.ts:24]: Already handled as of `0d89f56` (the commit this review is on) — went with your option 2 (JSON.stringify). `generateInitCode` emits `.domain(${kotlinStringLiteral(props.domain)})`, where:

```ts
kotlinStringLiteral = (v: string) => JSON.stringify(v).replace(/\$/g, '\\$')
```

JSON.stringify covers quotes, backslashes, and control chars (your `my"shop"` / backslash / newline cases); the extra `.replace` escapes `$`, which Kotlin treats as string-template interpolation and JSON escaping alone would miss. A unit test covers quotes, backslashes, and `$`.

I intentionally did **not** take the regex-validation option — we don't want to constrain the domain's character set at the plugin layer, so escaping (rather than rejecting) is the safer default. Let me know if you'd prefer validation on top.

File: package.json (hunk 7 above)
Comment (msouza-attentive) [reply package.json]: Fixed in `8648a27`: renamed `prepack` → `prepare` (`bob build && npm run build:plugin`).

One precise-mechanism note for the record: per the npm v10/v11 scripts docs, `prepack` actually *does* run "when installing a git dependency" — but the operative detail is that npm only installs a git dependency's **devDependencies** (here `bob` and `tsc`) and runs the build-from-source lifecycle **when a `prepare` script is present**. With only `prepack`, the git checkout's devDeps were never installed, so the build couldn't run and `plugin/build/` (and `lib/`) stayed absent → Expo failed at `require('./plugin/build')`. `prepare` is the hook that triggers the whole build-from-git path, which is also why `react-native-builder-bob` scaffolds `prepare` by default. `prepare` still runs on `npm pack`/`npm publish`, so the registry/tarball path is unchanged.

Verified with a real `git+file://…#feature/MSDK-404-expo-plugin` install into a clean app: pre-fix `plugin/build/` absent + prebuild crash; post-fix built and prebuild injects. Good catch — this also fixed the pre-existing `lib/` gap for git installs.

**Review by usharif-attentive — COMMENTED** [review]
Two findings from review, posted inline: README rendering regressions from the prettier reformat, and a double-init edge case on the plugin's manual-integration skip path.

File: README.md [hunk 8]
```diff
@@ -11,24 +12,23 @@ This project uses **npm** as the preferred package manager for consistency and a
...
 
 ## Installation
 
 Run `npm install @attentive-mobile/attentive-react-native-sdk` from your app's root directory.
 
 ## Setup with an AI coding agent
 
-> [!WARNING]
-> **Experimental.** Agent-assisted setup is a new, experimental feature. The [`AGENTS.md`](./AGENTS.md) guide and this flow may change, and results can vary by agent and project. Review whatever the agent changes before committing, and fall back to the manual steps below if anything looks off.
+> [!WARNING] > **Experimental.** Agent-assisted setup is a new, experimental feature. The [`AGENTS.md`](./AGENTS.md) guide and this flow may change, and results can vary by agent and project. Review whatever the agent changes before committing, and fall back to the manual steps below if anything looks off.
```
Comment (usharif-attentive) [inline README.md]: The prettier reformat collapsed the GitHub alert onto one line. `[!WARNING]` must be on its own blockquote line for GitHub to render it as an admonition — as written this now renders as literal text. Should be:

```markdown
> [!WARNING]
> **Experimental.** Agent-assisted setup is a new, experimental feature. ...
```

(More broadly: the mechanical README reformat is ~400 lines of churn unrelated to this PR — consider reverting it and keeping only the new Expo content, or doing it as a separate PR with settings that don't break the markdown.)

File: README.md [hunk 9]
```diff
@@ -39,15 +39,21 @@ If you use Claude Code, Cursor, Copilot, Codex, or another AI coding agent, you
 The agent flow intentionally stops at base integration. It does **not** wire up identify/clearUser, event recording, Creatives, marketing subscriptions, or push notifications — see the sections below for those.
 
 ## Usage
+
 See the [Bonni example app](https://github.com/attentive-mobile/attentive-react-native-sdk/blob/main/Bonni)
 for a sample of how the Attentive React Native SDK is used.
 
-__*** NOTE: Please refrain from using any private or undocumented classes or methods as they may change between releases. ***__
+****_ NOTE: Please refrain from using any private or undocumented classes or methods as they may change between releases. _****
```
Comment (usharif-attentive) [inline README.md]: The reformat mangled this emphasis: `****_ ... _****` is not valid Markdown and will render with literal asterisks. The original was `__*** NOTE ... ***__`.

File: README.md [hunk 10]
```diff
@@ -174,11 +184,61 @@ After the native initialization, all other SDK operations (`identify`, `recordAd
...
+**Notes:**
+
+- **Requires Expo SDK 50+** (Kotlin `MainApplication` templates). On older SDKs the plugin fails prebuild with an explicit error rather than generating broken code.
+- **iOS is untouched by the plugin.** iOS still initializes from TypeScript — the [`initialize()` call](#ios--initialize-from-typescript) is required either way.
+- The debug overlay (`enableDebugger`, see [Debugging Features](#debugging-features)) is wired only by the TypeScript `initialize()` call. The plugin's `mode: 'debug'` sets the _native_ SDK mode (creative debug view, verbose native logging) — a different switch.
+- If your `MainApplication` already contains a manual Attentive integration, the plugin leaves the file alone and prints a warning during prebuild — remove the manual code to let the plugin manage initialization.
+- Builder options beyond `domain`/`mode` (`notificationIconId`, `skipFatigueOnCreatives`, `logLevel`) are not yet exposed as plugin props.
+- Bare React Native apps (no prebuild) should not use the plugin — follow the manual instructions above.
+
+### Destroy the creative
```
Comment (usharif-attentive) [inline README.md]: This "Destroy the creative" section is a duplicate — the same section (with the same `destroyCreative()` snippet) still exists further down under the creative-triggering docs (line ~301). Looks like an accidental paste during the reformat; this copy should be removed.

File: plugin/src/withAndroid.ts [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+  const { contents, language } = modResults
+
+  if (hasManualAttentiveInit(contents)) {
+    console.warn(
+      '[attentive-react-native-sdk] Found an existing Attentive integration ' +
+        'in MainApplication that was not generated by this plugin. Skipping ' +
+        'code injection — remove the manual initialization to let the plugin ' +
+        'manage it.'
+    )
+    return contents
```
Comment (usharif-attentive) [inline plugin/src/withAndroid.ts]: Edge case: when manual Attentive code is detected **alongside existing plugin-generated blocks** (the `manual integration alongside generated blocks` test pins this), returning `contents` unchanged leaves *both* inits in the file — the app initializes twice, with potentially different domains/modes, until the developer intervenes. That's the failure mode this skip is meant to prevent.

Since the `@generated` blocks are plugin-owned (not app-owned code), consider stripping them on this path:

```ts
return stripGeneratedBlocks(contents)
```

That keeps the "never delete app-owned code" guarantee while leaving exactly one initialization. At minimum, the warning should mention that a previously generated block is still present in the file.

Comment (msouza-attentive) [reply plugin/src/withAndroid.ts]: Good catch — this was the one edge the skip path got wrong. Fixed in 2593f27 with your suggested approach: on the skip path we now `return stripGeneratedBlocks(contents)` instead of returning the file unchanged, so when a prior plugin-generated block coexists with a manual init, our block is removed and exactly the app's one `AttentiveSdk.initialize(...)` remains. It only ever strips plugin-owned `@generated` blocks, never app-owned code, so the guarantee holds.

Behavior by case:
- manual init only (no generated block) → strip is a no-op → file unchanged (as before)
- manual init **+** generated block → our block stripped → single init (the double-init you flagged)
- generated block only (no manual) → doesn't hit this path; `mergeContents` updates it in place as before

Also reworded the warning to say when a previously-generated block was removed. Updated the `manual integration alongside generated blocks` test to assert the generated block is gone, the manual init survives, and there's exactly one `initialize(` call (14/14 plugin tests pass).

Separately, re: the Markdown/formatting noise on this PR — that was editor format-on-save running Prettier over `.md` (Prettier 2.x mangles GitHub alerts like `> [!WARNING]`). Opened #101 to add a `.prettierignore` excluding Markdown so it stops happening repo-wide.
